### PR TITLE
WIP: Addition of generic type RNG

### PR DESCRIPTION
**Pull request by RustyPotato**
So, I am writing an integer version of rng. All over the code Clion compains about how longs might not fit into the ints in various places, so instead of casting everything I decided it'd be best to just have a function that returns ints. It follows the same format as rng_float. As for actually generating random numbers ... I just copied the long one we currently have and replaced longs with ints (really open to suggestions here). 

I'll be making my way through the files this week I hope, but I'll probably be a bit. So, opening this early as a WIP and open for comments.

### CODE REVIEW

**Comment by BevapDin** [comment]
Instead of a adding a new function name, reuse the existing one and overload it based on the parameters:
`int rng(int, int)` and `long rng(long, long)` - calling `rng` with `long` means one wants to get `long` back and calling it with `int` means one wants `int` back. You could even make it completely generic by making it a function template like `std::max`.

**Comment by BevapDin** [comment]
> want to change rng_double

Yes. This would be good.

> In the case someone called rng(int, long), which data type would you like returned?

You could automatically determine a type that fits all values from either (in this case a type that fits all `long` and all `int` values). But that is not easy and sometimes not possible, or might lead to surprising types (e.g. `rng(int, float)` would have to return a `double` if `int` and `float` have the same size).

The easy solution is to require the parameters to be of the same type. `std::max` does so, calling `std::max( 1, 1l )` will cause a compile time error. This makes the resulting type easy to determine (when reading the function call).

**Comment by AMurkin** [comment]
Some files have whitespace-only changes. And some rng calls not changed whitespace-wise.

File: src/iuse.cpp [hunk 1]
```diff
@@ -3240,35 +3240,38 @@ int iuse::granade_act(player *, item *it, bool t, const tripoint &pos)
             case 2:
                 sounds::sound(pos, 100, _("BUFFS!!"));
                 g->draw_explosion( pos, explosion_radius, c_green );
-                for( const tripoint &dest : g->m.points_in_radius( pos, explosion_radius ) ) {
```
Comment (kevingranade) [inline src/iuse.cpp]: You need a merge or rebase, this reverts part of a recently merged PR.
Just merge and take the version in HEAD for this hunk of code.

File: src/rng.cpp [hunk 2]
```diff
@@ -8,18 +8,12 @@
 #define _USE_MATH_DEFINES
 #include <cmath>
 
-long rng( long val1, long val2 )
+template<class T>
+const T& rng(const T& val1, const T& val2)
```
Comment (kevingranade) [inline src/rng.cpp]: This needs to be limited to numeric types as @BevapDin mentioned.
See http://en.cppreference.com/w/cpp/types/enable_if for the feature, and
https://github.com/CleverRaven/Cataclysm-DDA/blob/4589d3bddb94359cdf9c200bcd5fe44d8fcf0481/src/common_types.h#L13 for an example where it is used.

File: src/rng.cpp [hunk 3]
```diff
@@ -8,18 +8,13 @@
 #define _USE_MATH_DEFINES
 #include <cmath>
 
-long rng( long val1, long val2 )
+//template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+template <class T>
+const T rng(const T& val1, const T& val2)
```
Comment (ZhilkinSerg) [inline src/rng.cpp]: File need astyle.

Should be:
```C++
const T rng( const T &val1, const T &val2 ) 
```

File: src/rng.cpp [hunk 4]
```diff
@@ -30,7 +25,7 @@ bool one_in( int chance )
 //this works just like one_in, but it accepts doubles as input to calculate chances like "1 in 350,52"
 bool one_in_improved( double chance )
 {
-    return ( chance <= 1 || rng_float( 0, chance ) < 1 );
+    return ( chance <= 1 || rng( double(0), chance ) < 1 );
```
Comment (ZhilkinSerg) [inline src/rng.cpp]: Should be:
```C++
return ( chance <= 1 || rng( double( 0 ), chance ) < 1 );
```

File: src/rng.h [hunk 5]
```diff
@@ -8,8 +8,10 @@
 #include <functional>
 #include <array>
 
-long rng( long val1, long val2 );
-double rng_float( double val1, double val2 );
+//template<typename T, typename          = typename std::enable_if  <std::is_arithmetic<T>::value, T>::type>
+//template<typename T, typename TEnabled = typename std::enable_if_t<std::is_arithmetic<T>::value, T>>
+template<typename T>
+const T rng(const T& val1, const T& val2);
```
Comment (ZhilkinSerg) [inline src/rng.h]: File needs astyle too:

Should be:
```C++
const T rng( const T &val1, const T &val2 );
```

File: src/rng.h [hunk 6]
```diff
@@ -49,7 +51,7 @@ inline V random_entry( const C &container, D default_value )
         return default_value;
     }
     auto iter = container.begin();
-    std::advance( iter, rng( 0, container.size() - 1 ) );
+    std::advance( iter, rng( size_t(0), container.size() - 1 ) );
```
Comment (ZhilkinSerg) [inline src/rng.h]: Should be:
```C++
std::advance( iter, rng( size_t( 0 ), container.size() - 1 ) );
```

File: src/rng.h [hunk 7]
```diff
@@ -79,7 +81,7 @@ inline V random_entry( const C &container )
         return V();
     }
     auto iter = container.begin();
-    std::advance( iter, rng( 0, container.size() - 1 ) );
+    std::advance( iter, rng( size_t(0), container.size() - 1 ) );
```
Comment (ZhilkinSerg) [inline src/rng.h]: Should be:
```C++
std::advance( iter, rng( size_t( 0 ), container.size() - 1 ) );
```

File: src/rng.h [hunk 8]
```diff
@@ -110,7 +112,7 @@ inline typename std::enable_if < !is_std_array<C>::value,
         return default_value;
     }
     auto iter = container.begin();
-    std::advance( iter, rng( 0, container.size() - 1 ) );
+    std::advance( iter, rng( size_t(0), container.size() - 1 ) );
```
Comment (ZhilkinSerg) [inline src/rng.h]: Should be:
```C++
std::advance( iter, rng( size_t( 0 ), container.size() - 1 ) );
```

File: src/rng.h [hunk 9]
```diff
@@ -127,7 +129,7 @@ template<typename C, typename V = typename C::value_type>
 inline V random_entry_removed( C &container )
 {
     auto iter = container.begin();
-    std::advance( iter, rng( 0, container.size() - 1 ) );
+    std::advance( iter, rng( size_t(0), container.size() - 1 ) );
```
Comment (ZhilkinSerg) [inline src/rng.h]: Should be:
```C++
std::advance( iter, rng( size_t( 0 ), container.size() - 1 ) );
```

File: src/sounds.cpp [hunk 10]
```diff
@@ -131,7 +131,7 @@ static std::vector<centroid> cluster_sounds( std::vector<std::pair<tripoint, int
     const size_t max_map_distance = rl_dist( 0, 0, MAPSIZE * SEEX, MAPSIZE * SEEY );
     // Randomly choose cluster seeds.
     for( size_t i = recent_sounds.size(); i > stopping_point; i-- ) {
-        size_t index = rng( 0, i - 1 );
+        size_t index = rng( size_t(0), i - 1 );
```
Comment (ZhilkinSerg) [inline src/sounds.cpp]: File needs astyle.

Should be:
```C++
size_t index = rng( size_t( 0 ), i - 1 );
```

File: src/weather.cpp [hunk 11]
```diff
@@ -220,7 +220,7 @@ void item::add_rain_to_container(bool acid, int charges)
                 long total_poison = liq.poison * (orig) + (5 * added);
                 liq.poison = total_poison / liq.charges;
                 long leftover_poison = total_poison - liq.poison * liq.charges;
-                if (leftover_poison > rng(0, liq.charges)) {
+                if (leftover_poison > rng(0l, liq.charges)) {
```
Comment (ZhilkinSerg) [inline src/weather.cpp]: File needs astyle.

File needs astyle.

Should be:
```C++
if( leftover_poison > rng( 0l, liq.charges ) ) {
```

File: src/name.cpp [hunk 12]
```diff
@@ -105,7 +105,7 @@ std::string get( nameFlags searchFlags )
         }
 
         // make random selection and return result.
-        size_t choice = rng( 0, nChoices - 1 );
+        size_t choice = rng( size_t(0), nChoices - 1 );
```
Comment (ZhilkinSerg) [inline src/name.cpp]: Astyle like this:

```C++
size_t choice = rng( size_t( 0 ), nChoices - 1 );
```

File: src/rng.h [hunk 13]
```diff
@@ -8,8 +8,14 @@
 #include <functional>
 #include <array>
 
-long rng( long val1, long val2 );
-double rng_float( double val1, double val2 );
+template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+const T rng( const T& val1, const T& val2 )
```
Comment (ZhilkinSerg) [inline src/rng.h]:
```C++
const T rng( const T &val1, const T &val2 )
```

File: src/mattack_actors.cpp [hunk 14]
```diff
@@ -36,7 +36,7 @@ bool is_adjacent( const monster &z, const Creature &target )
 bool dodge_check( float max_accuracy, Creature &target )
 {
     ///\EFFECT_DODGE increases chance of dodging special attacks of monsters
-    float dodge = std::max( target.get_dodge() - rng( 0, max_accuracy ), 0.0f );
+    float dodge = std::max( target.get_dodge() - rng( float(0), max_accuracy ), 0.0f );
```
Comment (ZhilkinSerg) [inline src/mattack_actors.cpp]:
```C++
float dodge = std::max( target.get_dodge() - rng( float( 0 ), max_accuracy ), 0.0f );
```

**Comment by kevingranade** [comment]
> The key difference being the +1 or nothing. I am not sure how to proceed from here. Would it be best to add the +1 and hope the tests pass? Or to alter the tests to match the code we've written (sounds wrong).

Just split it back out into two seperate functions.
I don't remember the details, but this is dealing with rounding and needs to be present.

**Pull request closed without merging**